File: src/Day3.hs
module Day3
  (Coordinate(..),
   Wire,
   Distance,
   Direction (..),
   Segment(..),
   Coord(..),
   addAll,
   makeWire',
   Sign (..),
   extendWire,
   wholeWire,
   intersections,
   manhattendist,
   readDirection,
   readDistance,
   readSegment,
   day3a
   
  )
  where
import Data.List
import Data.List.Split
import Control.Monad
import Data.String.Utils (rstrip)

-- some Types

type Coordinate = (Int,Int)


data Coord = X|Y

type Wire = [Coordinate]

type Distance = Int

data Direction = R  |L |U |D
  deriving (Show, Eq, Read)

data Sign = Minus| Plus

type Segment = (Direction,Distance)



extendWire :: Wire -> Segment -> Wire
extendWire [] seg = makeWire' (0,0) seg
extendWire wire seg = makeWire' (head wire) seg

makeWire' :: Coordinate -> Segment-> Wire
makeWire' point (d,dist)
  |d == R = addAll Plus X dist point
  |d == L = addAll Minus X dist point
  |d == U = addAll Plus Y dist point
  |d == D = addAll Minus Y dist point

addAll :: Sign-> Coord -> Distance -> Coordinate -> Wire
addAll Plus X dist (x0,y0) = reverse [ (x,y0)|x<-[x0..(x0 + dist)]]
addAll Minus X dist (x0,y0) = [ (x,y0)|x<-[(x0 - dist)..x0]]
addAll Plus Y dist (x0,y0) = reverse [ (x0,y)|y<-[y0..(y0+dist)]]
addAll Minus Y dist (x0,y0) = [(x0,y)|y<-[(y0 - dist)..y0]]

wholeWire :: [Segment] -> Wire
wholeWire seglist = join $ scanl extendWire [(0,0)] seglist 

intersections :: Wire -> Wire -> Wire
intersections wire1 wire2 = filter (/= (0,0)) $ nub $ intersect wire1 wire2


manhattendist :: Coordinate -> Int
manhattendist (x,y) = abs x+abs y


mindist intersectionlist = minimum $ map manhattendist intersectionlist

day3a = do
  input3 <- readFile "./input/day3.txt"
  let wires = map (map readSegment) $ map (map (splitAt 1)) $ map (splitOn ",") $ splitOn "\n" $ rstrip input3
  let wire1 = wholeWire $ wires!!0
  let wire2 = wholeWire $ wires!!1
  let result = mindist $ intersections wire1 wire2
  print result

readDirection :: String -> Direction
readDirection = read

readDistance :: String ->Distance
readDistance = read

readSegment :: (String,String) -> Segment
readSegment (a,b)= (readDirection a, readDistance b)

File: rust/src/day3.rs
pub mod day3a{
    use std::iter::FromIterator;
    extern crate array_tool;
    use array_tool::vec::*;
    use std::fs;
    use std::str::FromStr;


    #[derive(Debug,PartialEq,Clone)]
    struct Coordinate (i32,i32);
    #[derive(Debug)]
    enum Direction {
        R,
        L,
        U,
        D
    }
    #[derive(Debug)]
    struct Segment (Direction,Distance);
    
    type Distance = i32;
    
    type Wire =Vec<Coordinate>;
    #[derive(Debug)]
    struct ParseDirectionError { /* fields omitted */ }
    
    impl FromStr for Direction {
        type Err = ParseDirectionError;
    
        
        #[inline]
        fn from_str(s: &str) -> Result<Direction, ParseDirectionError> {
            match s {
                "D"  => Ok(Direction::D),
                "R" => Ok(Direction::R),
                "L" => Ok(Direction::L),
                "U" => Ok(Direction::U),
                _   => Err(ParseDirectionError {/*fields*/}),
            }
        }
    }








    
    
    fn make_wire(coord: Coordinate, seg: &Segment ) -> Wire
    {   
        match seg {
           Segment (Direction::R ,a)=> push_to_x(coord.0,coord.0+a,coord.1),
    
           Segment (Direction::L,a)=>rev_push_to_x(coord.0-a,coord.0,coord.1),
            
           Segment (Direction::D,a)=>rev_push_to_y(coord.1-a,coord.1,coord.0),
           Segment (Direction::U,a)=>push_to_y(coord.1,coord.1+a,coord.0)
        
        
        }
    }
    fn extend_wire(seg: &Segment, mut wire: Wire) -> Wire
    {
        
        let maybelast: Option<Coordinate> = wire.pop();
        let last:Coordinate = check_maybe_coord(maybelast);
        let newWire : Wire = make_wire(last, seg);
        
        let resultvec: Wire = wire.into_iter().chain(newWire).collect();
        return resultvec


    }
    fn check_maybe_coord(opt:Option<Coordinate>)->Coordinate
    {
        match opt {
            None => Coordinate(0,0),
            Some(c)=>c
        }
    }
    
    fn push_to_x(startval:i32, endval:i32,yval:i32)->Vec<Coordinate>
    {
        let mut vector :Vec<Coordinate>= Vec::new();
        let mut index = startval;
        while index<(endval+1) {
            vector.push(Coordinate(index,yval));
            index+=1;
        }
        return vector
    }
    fn push_to_y(startval:i32, endval:i32,xval:i32)->Vec<Coordinate>
    {
        let mut vector :Vec<Coordinate>= Vec::new();
        let mut index = startval;
        while index<(endval+1) {
            vector.push(Coordinate(xval,index));
            index+=1;
        }
        return vector
    }
    fn rev_push_to_x(startval:i32,endval:i32,yval:i32)->Vec<Coordinate>
    {
        let mut vector :Vec<Coordinate> = push_to_x(startval, endval, yval);
        let oldwire = vector.into_iter();
        let reversed :Vec<Coordinate>= oldwire.rev().collect();
        return reversed

    }

    fn rev_push_to_y(startval:i32,endval:i32,xval:i32)->Vec<Coordinate>
    {
        let mut vector :Vec<Coordinate> = push_to_y(startval, endval, xval);
        let oldwire = vector.into_iter();
        let reversed :Vec<Coordinate>= oldwire.rev().collect();
        return reversed

    }
    fn whole_wire(seglist:&Vec<Segment>)->Wire{
        let mut v: Wire= vec![Coordinate (0,0)];
        let wire = seglist.iter().fold(v, |acc, x| extend_wire(x,acc));
        return wire

    }
    fn manhattendistance(coord:&Coordinate)->i32{
        coord.0.abs() +coord.1.abs()
    }
    fn car_cdr(s: &str) -> (&str, &str) {
        match s.chars().next() {
            Some(c) => s.split_at(c.len_utf8()),
            None => s.split_at(0),
        }
    }
    pub fn day3a_sol(){
        
        let filename = "../../input/day3.txt";
        let contents = fs::read_to_string(filename);
    
    
    
        let contents2= contents.unwrap();
        let thelines = contents2.lines();
        

        let mut seglists:Vec<Vec<Segment>>=Vec::new();
        for s in thelines {
            let mut segmentlist :Vec<Segment>= Vec::new();
            
            let segs: Vec<&str> = s.split(',').collect();
            for seg in segs {
                let (first_char, remainder) = car_cdr(seg);
                let dir:Direction = Direction::from_str(first_char).unwrap();
                let dist:Distance = i32::from_str(remainder).unwrap();
                let segment = Segment(dir,dist);
                segmentlist.push(segment);
                
            }
            seglists.push(segmentlist);
            


        }
        
        let wholewire1=whole_wire(&seglists[0]);
        let wholewire2=whole_wire(&seglists[1]);
        let intersections=wholewire1.intersect(wholewire2);
        
        let manhattendistances:Vec<i32>= intersections.iter()
                                                      .filter(|y| **y!= Coordinate(0,0))
                                                      .map( |x| manhattendistance(x)).collect();
        let mindist = manhattendistances.iter().min();
        match mindist {
            Some(min) => println!( "Min value: {}", min ),
            None      => println!( "Vector is empty" ),
        }

    }
}